File: backend/app/connectors/service.py
"""
Connectors Module — Service

Data ingestion logic for CSV and JSON files. Routers stay thin; all
parsing, normalization and inserts live here. Honors import settings
(date format, default currency) from the settings store.
"""
import csv
import json
import logging
import uuid
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.finance.models import Expense, ExpenseCategory, Revenue, RevenueSource
from app.properties.models import Property, PropertyType
from app.reservations.models import Reservation, ReservationSource, ReservationStatus

log = logging.getLogger("hostwise.connectors")

DATE_FORMATS = {
    "DD/MM/YYYY": "%d/%m/%Y",
    "MM/DD/YYYY": "%m/%d/%Y",
    "YYYY-MM-DD": "%Y-%m-%d",
}


def _parse_date(value: str, fmt: str) -> object:
    """Parse a date string honoring the configured import date format."""
    value = (value or "").strip()
    if not value:
        raise ValueError("empty date")
    for candidate in [DATE_FORMATS.get(fmt), "%Y-%m-%d", "%d/%m/%Y", "%m/%d/%Y"]:
        if not candidate:
            continue
        try:
            return datetime.strptime(value, candidate).date()
        except ValueError:
            continue
    raise ValueError(f"unparseable date: {value}")


class ConnectorService:
    """Handles file preview and import for CSV and JSON data."""

    def __init__(self, session: AsyncSession):
        self.session = session
        self._props_created = 0

    @staticmethod
    def read_file(file_path: str) -> tuple[str, list[str], list[dict]]:
        """Read a file, returning (format, columns, rows)."""
        if file_path.lower().endswith(".json"):
            with open(file_path, encoding="utf-8") as f:
                payload = json.load(f)
            if isinstance(payload, dict) and isinstance(payload.get("rows"), list):
                rows = payload["rows"]
            elif isinstance(payload, list):
                rows = payload
            else:
                raise ValueError("JSON must be an array of objects or {type, rows}")
            columns = list(rows[0].keys()) if rows else []
            return "json", columns, rows

        with open(file_path, encoding="utf-8-sig") as f:
            reader = csv.DictReader(f)
            columns = [c.lower().strip() for c in (reader.fieldnames or [])]
            rows = list(reader)
        return "csv", columns, rows

    @staticmethod
    def detect_type(import_type: str, columns: list[str]) -> str:
        if import_type != "auto":
            return import_type
        cols = " ".join(columns)
        if ("gross_revenue" in cols or "gross_amount" in cols) and "reservation_id" in cols:
            return "revenues"
        if "check_in" in cols and "check_out" in cols:
            return "reservations"
        if ("expense" in cols or "category" in cols) and "amount" in cols:
            return "expenses"
        return "reservations"

    async def _resolve_property(
        self,
        prop_map: dict,
        existing_props: dict,
        csv_pid: str,
        prop_name: str = "",
        city: str = "",
        country: str = "",
    ) -> uuid.UUID | None:
        csv_pid = (csv_pid or "").strip()
        prop_name = (prop_name or "").strip()

        if csv_pid and csv_pid in prop_map:
            return prop_map[csv_pid]
        if prop_name and prop_name.lower() in existing_props:
            prop_map[csv_pid] = existing_props[prop_name.lower()].id
            return prop_map[csv_pid]

        prop = Property(
            name=prop_name or f"Imported {csv_pid[:8] or uuid.uuid4().hex[:6]}",
            city=city or None,
            country=country or None,
            type=PropertyType.OTHER,
            bedrooms=1, bathrooms=1, max_guests=2,
        )
        self.session.add(prop)
        await self.session.flush()
        prop_map[csv_pid] = prop.id
        existing_props[prop.name.lower()] = prop
        self._props_created += 1
        return prop.id

    async def import_file(self, file_path: str, import_type: str) -> dict:
        """Import a CSV/JSON file into the database."""
        from app.settings.service import SettingsService

        settings = await SettingsService(self.session).get_all()
        date_fmt = settings.get("import_date_format", "DD/MM/YYYY")
        currency = settings.get("default_currency", "EUR")

        fmt, columns, rows = self.read_file(file_path)
        if not rows:
            return {"import_type": import_type, "imported": 0, "properties_created": 0, "errors": ["File is empty."]}

        # Build property map from existing properties
        result = await self.session.execute(
            select(Property).where(Property.is_deleted == False)
        )
        existing_props = {p.name.lower(): p for p in result.scalars().all()}
        prop_map: dict[str, uuid.UUID] = {}

        async def resolve_property(row: dict) -> uuid.UUID | None:
            pid = await self._resolve_property(
                prop_map,
                existing_props,
                row.get("property_id", ""),
                row.get("property_name", row.get("Property Name", "")),
                row.get("city", ""),
                row.get("country", ""),
            )
            if pid and not prop_map.get(row.get("property_id", "")):
                prop_map[row.get("property_id", "")] = pid
            return pid

        # Expense category cache — CSV imports carry a category name, so we
        # find-or-create the matching ExpenseCategory and link it by id.
        exp_cat_cache: dict[str, ExpenseCategory] = {}
        result = await self.session.execute(
            select(ExpenseCategory).where(ExpenseCategory.is_deleted == False)
        )
        for cat in result.scalars().all():
            exp_cat_cache[cat.name.lower()] = cat

        async def resolve_expense_category(name: str) -> ExpenseCategory | None:
            key = (name or "").strip().lower()
            if not key:
                return None
            cat = exp_cat_cache.get(key)
            if not cat:
                cat = ExpenseCategory(name=name.strip(), is_default=False, sort_order=0)
                self.session.add(cat)
                await self.session.flush()
                exp_cat_cache[key] = cat
            return cat

        detected = self.detect_type(import_type, columns)
        imported = 0
        errors: list[str] = []

        if detected == "reservations":
            for row in rows:
                try:
                    prop_id = await resolve_property(row)
                    if not prop_id:
                        continue
                    check_in = _parse_date(row.get("check_in") or row.get("Check-in") or "", date_fmt)
                    check_out = _parse_date(row.get("check_out") or row.get("Check-out") or "", date_fmt)
                    nights = int(row.get("nights", row.get("Nights", 0)) or 0) or (check_out - check_in).days
                    gross = float(row.get("gross_amount", row.get("gross_revenue", row.get("Amount", 0)) or 0))
                    status_map = {
                        "Confirmed": ReservationStatus.CONFIRMED,
                        "Cancelled": ReservationStatus.CANCELLED,
                        "Completed": ReservationStatus.COMPLETED,
                    }
                    status = status_map.get(row.get("status", "Confirmed"), ReservationStatus.CONFIRMED)
                    self.session.add(Reservation(
                        property_id=prop_id,
                        confirmation_code=(row.get("reservation_id") or row.get("confirmation_code") or "").strip() or None,
                        status=status,
                        source=ReservationSource.CSV,
                        check_in=check_in,
                        check_out=check_out,
                        nights=nights,
                        guest_name=(row.get("guest_name") or "").strip() or None,
                        gross_revenue=gross,
                        net_revenue=gross * 0.85,
                        cleaning_fee=0,
                        platform_fee=gross * 0.15,
                        taxes=0,
                        currency=currency,
                        number_of_guests=2,
                        property_name=(row.get("property_name") or "").strip() or None,
                    ))
                    imported += 1
                except Exception as exc:  # noqa: BLE001 - per-row import errors are collected and surfaced
                    errors.append(str(exc))

        elif detected == "revenues":
            for row in rows:
                try:
                    prop_id = await resolve_property(row)
                    if not prop_id:
                        continue
                    rev_date = _parse_date(row.get("date") or "", date_fmt)
                    gross = float(row.get("gross_revenue", row.get("gross_amount", 0)) or 0)
                    comm = float(row.get("management_commission", row.get("commission_amount", 0)) or 0)
                    net = float(row.get("net_revenue", row.get("net_amount", gross - comm)) or 0)
                    source_map = {"airbnb": RevenueSource.AIRBNB, "booking": RevenueSource.BOOKING,
                                  "direct": RevenueSource.DIRECT, "csv": RevenueSource.CSV}
                    source = source_map.get((row.get("source", "csv") or "csv").strip().lower(), RevenueSource.CSV)
                    self.session.add(Revenue(
                        property_id=prop_id,
                        date=rev_date,
                        gross_amount=gross,
                        commission_amount=comm,
                        net_amount=net,
                        source=source,
                        currency=currency,
                    ))
                    imported += 1
                except Exception as exc:  # noqa: BLE001 - per-row import errors are collected and surfaced
                    errors.append(str(exc))

        elif detected == "expenses":
            for row in rows:
                try:
                    prop_id = await resolve_property(row)
                    if not prop_id:
                        continue
                    exp_date = _parse_date(row.get("date") or "", date_fmt)
                    amount = float(row.get("amount", 0) or 0)
                    cat_name = (row.get("category") or "").strip()
                    category = await resolve_expense_category(cat_name) if cat_name else None
                    self.session.add(Expense(
                        property_id=prop_id,
                        date=exp_date,
                        amount=amount,
                        currency=currency,
                        category_id=category.id if category else None,
                        description=cat_name or None,
                        vendor=(row.get("vendor") or "").strip() or None,
                    ))
                    imported += 1
                except Exception as exc:  # noqa: BLE001 - per-row import errors are collected and surfaced
                    errors.append(str(exc))
        else:
            errors.append(f"Unsupported import type: {detected}")

        return {
            "format": fmt,
            "import_type": detected,
            "imported": imported,
            "properties_created": self._props_created,
            "errors": errors[:10],
        }

File: backend-rs/src/connectors/service.rs
//! Connectors service: CSV/JSON preview + import, and iCal import.
//! Mirrors `backend/app/connectors/service.py`.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use chrono::NaiveDate;
use serde_json::{json, Value};
use sqlx::SqlitePool;
use uuid::Uuid;

use crate::connectors::ical;
use crate::core::error::AppError;
use crate::core::time::now_iso;
use crate::finance::models::{Expense, Revenue};
use crate::finance::repository as finance_repo;
use crate::properties::models::Property;
use crate::properties::repository as props_repo;
use crate::reservations::models::Reservation;
use crate::reservations::repository as res_repo;
use crate::settings::service as settings;

fn get<'a>(row: &'a HashMap<String, String>, keys: &[&str]) -> &'a str {
    for k in keys {
        if let Some(v) = row.get(*k) {
            if !v.trim().is_empty() {
                return v;
            }
        }
    }
    ""
}

fn read_bytes_lossy(path: &Path) -> Result<String, String> {
    let bytes = std::fs::read(path).map_err(|e| e.to_string())?;
    let content = String::from_utf8_lossy(&bytes).to_string();
    Ok(content
        .strip_prefix('\u{feff}')
        .unwrap_or(&content)
        .to_string())
}

/// Read a CSV file into (columns, rows of string values).
fn read_csv(
    path: &Path,
    delimiter: u8,
) -> Result<(Vec<String>, Vec<HashMap<String, String>>), String> {
    let content = read_bytes_lossy(path)?;
    let mut reader = csv::ReaderBuilder::new()
        .delimiter(delimiter)
        .has_headers(true)
        .from_reader(content.as_bytes());
    let headers: Vec<String> = reader
        .headers()
        .map_err(|e| e.to_string())?
        .iter()
        .map(|s| s.trim().to_lowercase())
        .collect();
    let mut rows = Vec::new();
    for rec in reader.records() {
        let rec = rec.map_err(|e| e.to_string())?;
        let mut map = HashMap::new();
        for (i, h) in headers.iter().enumerate() {
            map.insert(h.clone(), rec.get(i).unwrap_or("").to_string());
        }
        rows.push(map);
    }
    Ok((headers, rows))
}

/// Read a JSON file (array of objects or `{type, rows}`) into rows.
fn read_json(path: &Path) -> Result<(Vec<String>, Vec<HashMap<String, String>>), String> {
    let content = read_bytes_lossy(path)?;
    let v: Value = serde_json::from_str(&content).map_err(|e| e.to_string())?;
    let arr: Vec<Value> = match &v {
        Value::Array(a) => a.clone(),
        Value::Object(o) => o
            .get("rows")
            .and_then(|r| r.as_array())
            .cloned()
            .ok_or("JSON must be an array of objects or {type, rows}")?,
        _ => return Err("JSON must be an array of objects or {type, rows}".into()),
    };
    let mut cols: Vec<String> = Vec::new();
    for r in &arr {
        if let Value::Object(o) = r {
            for k in o.keys() {
                if !cols.contains(k) {
                    cols.push(k.clone());
                }
            }
        }
    }
    let mut rows = Vec::new();
    for r in &arr {
        if let Value::Object(o) = r {
            let mut map = HashMap::new();
            for (k, val) in o {
                map.insert(
                    k.clone(),
                    val.as_str()
                        .map(|s| s.to_string())
                        .unwrap_or_else(|| val.to_string()),
                );
            }
            rows.push(map);
        }
    }
    Ok((cols, rows))
}

/// Read a file, returning (format, columns, rows).
pub fn read_file(
    path: &Path,
    _encoding: Option<&str>,
    delimiter: Option<&str>,
) -> Result<(String, Vec<String>, Vec<HashMap<String, String>>), String> {
    let is_json = path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.eq_ignore_ascii_case("json"))
        .unwrap_or(false);
    if is_json {
        let (cols, rows) = read_json(path)?;
        return Ok(("json".into(), cols, rows));
    }
    let delim = delimiter.unwrap_or(",");
    let delim_byte = delim.as_bytes().first().copied().unwrap_or(b',');
    let (cols, rows) = read_csv(path, delim_byte)?;
    Ok(("csv".into(), cols, rows))
}

/// Auto-detect import type from column names.
pub fn detect_type(import_type: &str, columns: &[String]) -> String {
    if import_type != "auto" {
        return import_type.to_string();
    }
    let cols = columns.join(" ");
    if (cols.contains("gross_revenue") || cols.contains("gross_amount"))
        && cols.contains("reservation_id")
    {
        return "revenues".into();
    }
    if cols.contains("check_in") && cols.contains("check_out") {
        return "reservations".into();
    }
    if (cols.contains("expense") || cols.contains("category")) && cols.contains("amount") {
        return "expenses".into();
    }
    "reservations".into()
}

/// Parse a date honoring the configured import date format.
fn parse_date(value: &str, fmt: &str) -> Result<NaiveDate, String> {
    let v = value.trim();
    if v.is_empty() {
        return Err("empty date".into());
    }
    let candidates: &[&str] = match fmt {
        "DD/MM/YYYY" => &["%d/%m/%Y", "%Y-%m-%d", "%m/%d/%Y"],
        "MM/DD/YYYY" => &["%m/%d/%Y", "%Y-%m-%d", "%d/%m/%Y"],
        _ => &["%Y-%m-%d", "%d/%m/%Y", "%m/%d/%Y"],
    };
    for c in candidates {
        if let Ok(d) = NaiveDate::parse_from_str(v, c) {
            return Ok(d);
        }
    }
    Err(format!("unparseable date: {v}"))
}

/// Find or create a property; returns (id, created).
async fn resolve_property(
    pool: &SqlitePool,
    prop_map: &mut HashMap<String, String>,
    existing_props: &mut HashMap<String, String>,
    csv_pid: &str,
    prop_name: &str,
    city: &str,
    country: &str,
) -> Result<(Option<String>, bool), AppError> {
    let csv_pid = csv_pid.trim();
    let prop_name = prop_name.trim();

    if !csv_pid.is_empty() && prop_map.contains_key(csv_pid) {
        return Ok((Some(prop_map[csv_pid].clone()), false));
    }
    if !prop_name.is_empty() {
        let key = prop_name.to_lowercase();
        if let Some(id) = existing_props.get(&key) {
            prop_map.insert(csv_pid.to_string(), id.clone());
            return Ok((Some(id.clone()), false));
        }
    }

    let name = if prop_name.is_empty() {
        let suffix = if csv_pid.is_empty() {
            Uuid::new_v4().simple().to_string()[..6].to_string()
        } else {
            csv_pid.chars().take(8).collect()
        };
        format!("Imported {suffix}")
    } else {
        prop_name.to_string()
    };

    let now = now_iso();
    let id = Uuid::new_v4().to_string();
    let property = Property {
        id: id.clone(),
        name: name.clone(),
        r#type: "other".into(),
        status: "active".into(),
        address: None,
        city: if city.trim().is_empty() {
            None
        } else {
            Some(city.trim().to_string())
        },
        state: None,
        country: if country.trim().is_empty() {
            None
        } else {
            Some(country.trim().to_string())
        },
        postal_code: None,
        latitude: None,
        longitude: None,
        bedrooms: 1,
        bathrooms: 1.0,
        max_guests: 2,
        square_meters: None,
        acquisition_cost: None,
        monthly_mortgage: None,
        target_occupancy: None,
        target_annual_revenue: None,
        notes: None,
        deleted_at: None,
        created_at: now.clone(),
        updated_at: now,
        sync_id: None,
    };
    props_repo::insert(pool, &property).await?;
    prop_map.insert(csv_pid.to_string(), id.clone());
    existing_props.insert(name.to_lowercase(), id.clone());
    Ok((Some(id), true))
}

/// Find or create an expense category by name (case-insensitive).
async fn resolve_expense_category(
    pool: &SqlitePool,
    cache: &mut HashMap<String, String>,
    name: &str,
) -> Result<Option<String>, AppError> {
    let key = name.trim().to_lowercase();
    if key.is_empty() {
        return Ok(None);
    }
    if let Some(id) = cache.get(&key) {
        return Ok(Some(id.clone()));
    }
    let id = finance_repo::find_or_create_expense_category(pool, name).await?;
    cache.insert(key, id.clone());
    Ok(Some(id))
}

/// Import a previously-uploaded CSV/JSON file into the database.
pub async fn import_file(
    pool: &SqlitePool,
    file_path: &Path,
    import_type: &str,
) -> Result<Value, AppError> {
    let all = settings::get_all(pool).await?;
    let date_fmt = all["import_date_format"].as_str().unwrap_or("DD/MM/YYYY");
    let currency = all["default_currency"].as_str().unwrap_or("EUR");
    let enc = all["import_encoding"].as_str().unwrap_or("UTF-8");
    let delim = all["import_delimiter"].as_str().unwrap_or(",");

    let (fmt, columns, rows) =
        read_file(file_path, Some(enc), Some(delim)).map_err(AppError::Validation)?;
    if rows.is_empty() {
        return Ok(json!({
            "import_type": import_type, "imported": 0, "skipped": 0,
            "properties_created": 0, "errors": ["File is empty."],
        }));
    }

    let existing: Vec<(String, String)> =
        sqlx::query_as("SELECT id, name FROM properties WHERE is_deleted = 0")
            .fetch_all(pool)
            .await?;
    let mut existing_props: HashMap<String, String> = HashMap::new();
    for (id, name) in &existing {
        existing_props.insert(name.to_lowercase(), id.clone());
    }
    let mut prop_map: HashMap<String, String> = HashMap::new();
    let mut props_created: i64 = 0;

    let mut exp_cat_cache: HashMap<String, String> = HashMap::new();
    let existing_cats: Vec<(String, String)> =
        sqlx::query_as("SELECT id, name FROM expense_categories WHERE is_deleted = 0")
            .fetch_all(pool)
            .await?;
    for (id, name) in &existing_cats {
        exp_cat_cache.insert(name.to_lowercase(), id.clone());
    }

    let detected = detect_type(import_type, &columns);

    // Idempotency sets (natural keys, mirrors the Python service).
    let mut existing_res_codes: HashSet<String> = sqlx::query_scalar(
        "SELECT confirmation_code FROM reservations WHERE is_deleted = 0 AND confirmation_code IS NOT NULL",
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .collect();
    let mut existing_rev_keys: HashSet<String> = sqlx::query_as::<_, (String, String, f64, f64, String)>(
        "SELECT property_id, date, gross_amount, net_amount, source FROM revenues WHERE is_deleted = 0",
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .map(|(p, d, g, n, s)| format!("{p}|{d}|{g:.4}|{n:.4}|{s}"))
    .collect();
    let mut existing_exp_keys: HashSet<String> = sqlx::query_as::<_, (String, String, f64, String, String)>(
        "SELECT property_id, date, amount, COALESCE(vendor,''), COALESCE(description,'') FROM expenses WHERE is_deleted = 0",
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .map(|(p, d, a, v, c)| format!("{p}|{d}|{a:.4}|{v}|{c}"))
    .collect();

    let mut imported: i64 = 0;
    let mut skipped: i64 = 0;
    let mut errors: Vec<String> = Vec::new();

    if detected == "reservations" {
        for row in &rows {
            let outcome: Result<bool, AppError> = async {
                let (prop_id, created) = resolve_property(
                    pool,
                    &mut prop_map,
                    &mut existing_props,
                    get(row, &["property_id"]),
                    get(row, &["property_name", "Property Name"]),
                    get(row, &["city"]),
                    get(row, &["country"]),
                )
                .await?;
                let Some(prop_id) = prop_id else {
                    return Ok(false);
                };
                if created {
                    props_created += 1;
                }
                let code = get(row, &["reservation_id", "confirmation_code"]).to_string();
                if !code.is_empty() && existing_res_codes.contains(&code) {
                    return Ok(false);
                }
                let check_in = parse_date(get(row, &["check_in", "Check-in"]), date_fmt)
                    .map_err(AppError::Validation)?;
                let check_out = parse_date(get(row, &["check_out", "Check-out"]), date_fmt)
                    .map_err(AppError::Validation)?;
                let mut nights: i64 = get(row, &["nights", "Nights"]).trim().parse().unwrap_or(0);
                if nights == 0 {
                    nights = (check_out - check_in).num_days();
                }
                let gross: f64 = get(row, &["gross_amount", "gross_revenue", "Amount"])
                    .trim()
                    .parse()
                    .unwrap_or(0.0);
                let status = match get(row, &["status"]).trim() {
                    "Cancelled" => "cancelled",
                    "Completed" => "completed",
                    _ => "confirmed",
                };
                let now = now_iso();
                let reservation = Reservation {
                    id: Uuid::new_v4().to_string(),
                    property_id: prop_id.clone(),
                    listing_id: None,
                    external_id: None,
                    confirmation_code: if code.is_empty() {
                        None
                    } else {
                        Some(code.clone())
                    },
                    status: status.to_string(),
                    source: "csv".into(),
                    check_in: check_in.format("%Y-%m-%d").to_string(),
                    check_out: check_out.format("%Y-%m-%d").to_string(),
                    booked_at: None,
                    cancelled_at: None,
                    nights,
                    guest_name: {
                        let g = get(row, &["guest_name"]);
                        if g.trim().is_empty() {
                            None
                        } else {
                            Some(g.trim().to_string())
                        }
                    },
                    guest_email: None,
                    guest_phone: None,
                    number_of_guests: 2,
                    gross_revenue: gross,
                    cleaning_fee: 0.0,
                    platform_fee: gross * 0.15,
                    taxes: 0.0,
                    net_revenue: gross * 0.85,
                    currency: currency.to_string(),
                    property_name: {
                        let p = get(row, &["property_name"]);
                        if p.trim().is_empty() {
                            None
                        } else {
                            Some(p.trim().to_string())
                        }
                    },
                    property_city: None,
                    property_country: None,
                    notes: None,
                    deleted_at: None,
                    is_deleted: false,
                    created_at: now.clone(),
                    updated_at: now,
                    sync_id: None,
                };
                res_repo::insert(pool, &reservation).await?;
                if !code.is_empty() {
                    existing_res_codes.insert(code);
                }
                Ok(true)
            }
            .await;
            match outcome {
                Ok(true) => imported += 1,
                Ok(false) => skipped += 1,
                Err(e) => errors.push(e.to_string()),
            }
        }
    } else if detected == "revenues" {
        for row in &rows {
            let outcome: Result<bool, AppError> = async {
                let (prop_id, created) = resolve_property(
                    pool,
                    &mut prop_map,
                    &mut existing_props,
                    get(row, &["property_id"]),
                    get(row, &["property_name", "Property Name"]),
                    "",
                    "",
                )
                .await?;
                let Some(prop_id) = prop_id else {
                    return Ok(false);
                };
                if created {
                    props_created += 1;
                }
                let rev_date =
                    parse_date(get(row, &["date"]), date_fmt).map_err(AppError::Validation)?;
                let gross: f64 = get(row, &["gross_revenue", "gross_amount"])
                    .trim()
                    .parse()
                    .unwrap_or(0.0);
                let comm: f64 = get(row, &["management_commission", "commission_amount"])
                    .trim()
                    .parse()
                    .unwrap_or(0.0);
                let net: f64 = get(row, &["net_revenue", "net_amount"])
                    .trim()
                    .parse()
                    .unwrap_or(gross - comm);
                let source = match get(row, &["source"]).trim().to_lowercase().as_str() {
                    "airbnb" => "airbnb",
                    "booking" => "booking",
                    "direct" => "direct",
                    _ => "csv",
                };
                let date_str = rev_date.format("%Y-%m-%d").to_string();
                let key = format!("{prop_id}|{date_str}|{gross:.4}|{net:.4}|{source}");
                if existing_rev_keys.contains(&key) {
                    return Ok(false);
                }
                let now = now_iso();
                let revenue = Revenue {
                    id: Uuid::new_v4().to_string(),
                    property_id: prop_id.clone(),
                    reservation_id: None,
                    category_id: None,
                    date: date_str,
                    gross_amount: gross,
                    commission_amount: comm,
                    net_amount: net,
                    source: source.to_string(),
                    currency: currency.to_string(),
                    description: None,
                    notes: None,
                    deleted_at: None,
                    is_deleted: false,
                    created_at: now.clone(),
                    updated_at: now,
                    sync_id: None,
                };
                finance_repo::insert_revenue(pool, &revenue).await?;
                existing_rev_keys.insert(key);
                Ok(true)
            }
            .await;
            match outcome {
                Ok(true) => imported += 1,
                Ok(false) => skipped += 1,
                Err(e) => errors.push(e.to_string()),
            }
        }
    } else if detected == "expenses" {
        for row in &rows {
            let outcome: Result<bool, AppError> = async {
                let (prop_id, created) = resolve_property(
                    pool,
                    &mut prop_map,
                    &mut existing_props,
                    get(row, &["property_id"]),
                    get(row, &["property_name", "Property Name"]),
                    "",
                    "",
                )
                .await?;
                let Some(prop_id) = prop_id else {
                    return Ok(false);
                };
                if created {
                    props_created += 1;
                }
                let exp_date =
                    parse_date(get(row, &["date"]), date_fmt).map_err(AppError::Validation)?;
                let amount: f64 = get(row, &["amount"]).trim().parse().unwrap_or(0.0);
                let cat_name = get(row, &["category"]).trim().to_string();
                let vendor = get(row, &["vendor"]).trim().to_string();
                let date_str = exp_date.format("%Y-%m-%d").to_string();
                let key = format!("{prop_id}|{date_str}|{amount:.4}|{vendor}|{cat_name}");
                if existing_exp_keys.contains(&key) {
                    return Ok(false);
                }
                let category = if cat_name.is_empty() {
                    None
                } else {
                    resolve_expense_category(pool, &mut exp_cat_cache, &cat_name).await?
                };
                let now = now_iso();
                let expense = Expense {
                    id: Uuid::new_v4().to_string(),
                    property_id: prop_id.clone(),
                    category_id: category,
                    date: date_str,
                    amount,
                    currency: currency.to_string(),
                    vendor: if vendor.is_empty() {
                        None
                    } else {
                        Some(vendor.clone())
                    },
                    payment_method: None,
                    description: if cat_name.is_empty() {
                        None
                    } else {
                        Some(cat_name.clone())
                    },
                    notes: None,
                    is_recurring: false,
                    receipt_url: None,
                    deleted_at: None,
                    is_deleted: false,
                    created_at: now.clone(),
                    updated_at: now,
                    sync_id: None,
                };
                finance_repo::insert_expense(pool, &expense).await?;
                existing_exp_keys.insert(key);
                Ok(true)
            }
            .await;
            match outcome {
                Ok(true) => imported += 1,
                Ok(false) => skipped += 1,
                Err(e) => errors.push(e.to_string()),
            }
        }
    } else {
        errors.push(format!("Unsupported import type: {detected}"));
    }

    Ok(json!({
        "format": fmt,
        "import_type": detected,
        "imported": imported,
        "skipped": skipped,
        "properties_created": props_created,
        "errors": errors.iter().take(10).cloned().collect::<Vec<_>>(),
    }))
}

/// Import reservations from an .ics calendar export (idempotent by UID).
pub async fn import_ical(
    pool: &SqlitePool,
    file_path: &Path,
    property_id: &str,
) -> Result<Value, AppError> {
    let content = read_bytes_lossy(file_path).map_err(AppError::Validation)?;
    let events = ical::parse_ics(&content);

    if events.is_empty() {
        return Ok(json!({
            "format": "ics", "import_type": "ical", "imported": 0, "skipped": 0,
            "properties_created": 0, "errors": ["No VEVENT entries found in the calendar file."],
        }));
    }

    let prop = props_repo::get_by_id(pool, property_id).await?;
    let Some(prop) = prop else {
        return Ok(json!({
            "format": "ics", "import_type": "ical", "imported": 0, "skipped": 0,
            "properties_created": 0, "errors": ["Property not found."],
        }));
    };

    let all = settings::get_all(pool).await?;
    let currency = all["default_currency"].as_str().unwrap_or("EUR");

    let mut existing_uids: HashSet<String> = sqlx::query_scalar(
        "SELECT external_id FROM reservations WHERE external_id IS NOT NULL AND property_id = ?",
    )
    .bind(property_id)
    .fetch_all(pool)
    .await?
    .into_iter()
    .collect();

    let mut imported: i64 = 0;
    let mut skipped: i64 = 0;
    for ev in events {
        if !ev.uid.is_empty() && existing_uids.contains(&ev.uid) {
            skipped += 1;
            continue;
        }
        let now = now_iso();
        let reservation = Reservation {
            id: Uuid::new_v4().to_string(),
            property_id: property_id.to_string(),
            listing_id: None,
            external_id: Some(ev.uid.clone()),
            confirmation_code: Some(ev.uid.clone()),
            status: "confirmed".into(),
            source: "ical".into(),
            check_in: ev.check_in.format("%Y-%m-%d").to_string(),
            check_out: ev.check_out.format("%Y-%m-%d").to_string(),
            booked_at: None,
            cancelled_at: None,
            nights: ev.nights,
            guest_name: if ev.summary.trim().is_empty() {
                None
            } else {
                Some(ev.summary.clone())
            },
            guest_email: None,
            guest_phone: None,
            number_of_guests: 1,
            gross_revenue: 0.0,
            cleaning_fee: 0.0,
            platform_fee: 0.0,
            taxes: 0.0,
            net_revenue: 0.0,
            currency: currency.to_string(),
            property_name: Some(prop.name.clone()),
            property_city: None,
            property_country: None,
            notes: None,
            deleted_at: None,
            is_deleted: false,
            created_at: now.clone(),
            updated_at: now,
            sync_id: None,
        };
        res_repo::insert(pool, &reservation).await?;
        existing_uids.insert(ev.uid);
        imported += 1;
    }

    Ok(json!({
        "format": "ics",
        "import_type": "ical",
        "imported": imported,
        "skipped": skipped,
        "properties_created": 0,
        "errors": [],
    }))
}
